src/Database/Expression/CommonTableExpression.php

```diff
@@ -0,0 +1,297 @@
+                    Debugger::getType($field),
+                    $index
+                ));
+            }
+        }
+
+        $this->fields = $fields;
+
+        return $this;
+    }
+
+    /**
+     * Returns the modifiers to use for the CTE.
+     *
+     * @return \Cake\Database\ExpressionInterface[]|string[]
+     */
+    public function getModifiers(): array
+    {
+        return $this->modifiers;
+    }
+
+    /**
+     * Sets the modifiers to use for the CTE.
+     *
+     * @param \Cake\Database\ExpressionInterface[]|string[] $modifiers The modifiers to use for the CTE.
+     * @return $this
+     */
+    public function setModifiers(array $modifiers)
+    {
+        foreach ($modifiers as $index => $modifier) {
+            if (
+                !($modifier instanceof ExpressionInterface) &&
+                !is_string($modifier)
+            ) {
+                throw new InvalidArgumentException(sprintf(
+                    'The `$modifiers` argument must contain only instances of `%s`, or strings, ' .
+                        '`%s` given at index `%d`.',
+                    ExpressionInterface::class,
+                    Debugger::getType($modifier),
+                    $index
+                ));
+            }
+        }
+
+        $this->modifiers = $modifiers;
+
+        return $this;
+    }
+
+    /**
+     * Returns the CTE query definition.
+     *
+     * @return \Cake\Database\ExpressionInterface|string
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * Sets the CTE query definition.
+     *
+     * @param \Cake\Database\ExpressionInterface|string $query The CTE query definition.
+     * @return $this
+     */
+    public function setQuery($query)
+    {
+        if (
+            !($query instanceof ExpressionInterface) &&
+            !is_string($query)
+        ) {
+            throw new InvalidArgumentException(sprintf(
+                'The `$query` argument must be either an instance of `%s`, or a string, `%s` given.',
+                ExpressionInterface::class,
+                Debugger::getType($query)
+            ));
+        }
+
+        $this->query = $query;
+
+        return $this;
+    }
+
+    /**
+     * Returns whether the CTE operates recursively.
+     *
+     * @return bool
+     */
+    public function isRecursive(): bool
+    {
+        return $this->recursive;
+    }
+
+    /**
+     * Sets whether the CTE operates recursively.
+     *
+     * @param bool $recursive Indicates whether the CTE query operates recursively.
+     * @return $this
+     */
+    public function setRecursive(bool $recursive)
+    {
+        $this->recursive = $recursive;
+
+        return $this;
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param string $name The CTE name.
+     * @param \Cake\Database\ExpressionInterface|string $query The CTE query definition.
+     */
+    public function __construct(string $name, $query)
```

Comment:
Can you move the constructor at the top please.

Reply:
done